main.nf

```
@@ -68,14 +68,14 @@ workflow {
 
     // Use the UCSC FASTA and FAI files if provided, otherwise use the NCBI FASTA and FAI files
     def fasta = params.ucsc_fasta ? Channel.value([[id: 'fasta'], file(params.ucsc_fasta)]) :
-        params.ncbi_fasta ? Channel.value([[id: 'fasta'], file(params.ncbi_fasta)]) : [[:], []]
+        params.ncbi_fasta ? Channel.value([[id: 'fasta'], file(params.ncbi_fasta)]) : Channel.empty()
```

I found this solves the error descriped in #59 . Would you mind double check?

---

That's weird though, that channel should never be empty. It should contain one of the two fasta files. Are you sure `--ncbi_fasta` and/or `--ucsc_fasta` are set?

---

we shall allow user to put either one or both. The error happens when one is empty.

---

That's weird, it should check if the UCSC fasta exists and use that one if that is the case, otherwise it checks if the NCBI fasta exists and use that one if that is the case.

---

`fasta` was not the problem. It is not used actually so far. The error might because of these two parts

```
    def ucsc_fasta = params.ucsc_fasta ? Channel.value([[id: 'ucsc'], file(params.ucsc_fasta)]) : [[:], []]
    def ucsc_fai = params.ucsc_fai ? Channel.value([[id: 'ucsc'], file(params.ucsc_fai)]) : [[:], []]
    def ncbi_fasta = params.ncbi_fasta ? Channel.value([[id: 'ncbi'], file(params.ncbi_fasta)]) : [[:], []]
    def ncbi_fai = params.ncbi_fai ? Channel.value([[id: 'ncbi'], file(params.ncbi_fai)]) : [[:], []]
```
and 
```
    def ch_ucsc_refs = ucsc_fasta
        .join(ucsc_fai, failOnDuplicate:true, failOnMismatch:true)
        .join(ucsc_dict, failOnDuplicate:true, failOnMismatch:true)
    
    def ch_ncbi_refs = ncbi_fasta
        .join(ncbi_fai, failOnDuplicate:true, failOnMismatch:true)
        .join(ncbi_dict, failOnDuplicate:true, failOnMismatch:true) 
```
join raises error on [[:], []] as list

---

Ah yes that makes sense in that case! I would maybe set the default to `[[id:'ucsc'], []]` and `[[id:'ncbi'], []]`. That way the join will go through and give an error in a module, making it so that samples that depend on these references will fail instead of being skipped

---

I tried it still gives an error if I do not give params.ncbi_fasta
`Missing process or function join([[failOnDuplicate:true, failOnMismatch:true], [[id:ncbi], []]])`.

When you have time you can also play with the samplesheet and leave ncbi_fasta as empty.
For this PR i will leave it like now, if it is fine?

---

I'll take a quick look :p

---

I pushed a quick fix for the issue :)

---

The problem was the dictionary channels that were empty

---

it works! with defining dict channel and adding `Channel.of` everywhere! Thanks!